File: frontend/src/types/market.ts
export interface HealthStatus {
  status: string
  app_name: string
  mode: string
  scheduler_enabled: boolean
  market_db_path: string
  strategy_store: string
}

export interface MetricCard {
  label: string
  value: string
  change: string
  tone: 'positive' | 'negative' | 'neutral'
  description: string
}

export interface IndustryHeat {
  industry: string
  score: number
  momentum: string
}

export interface MarketPulsePoint {
  date: string
  score: number
  turnover: number
}

export interface DashboardSummary {
  headline: string
  updated_at: string
  market_context: MarketContextSummary
  benchmark_indices: BenchmarkIndex[]
  breadth_snapshot: MarketBreadthSnapshot
  market_capital_flow: MarketCapitalFlowOverview
  market_overview: MetricCard[]
  hot_industries: IndustryHeat[]
  market_pulse: MarketPulsePoint[]
  top_recommendations: RecommendationItem[]
  risk_flags: string[]
}

export interface MarketContextSummary {
  regime: 'risk_on' | 'balanced' | 'risk_off'
  regime_label: string
  summary: string
  action_hint: string
  watch_points: string[]
  metrics: MetricCard[]
}

export interface BenchmarkIndex {
  code: string
  name: string
  latest_price: number
  change_pct: number
  return_20d: number | null
  trend: string
  takeaway: string
}

export interface MarketBreadthSnapshot {
  scope_label: string
  total_count: number
  advancers: number
  decliners: number
  advance_ratio: number
  strong_count: number
  strong_ratio: number
  avg_change: number
  avg_turnover: number
  top_industry: string | null
  top_two_share: number
  limit_up_like: number | null
  limit_down_like: number | null
  summary: string
}

export interface MarketCapitalFlowOverview {
  status: 'ready' | 'derived' | 'placeholder'
  scope_label: string
  summary: string
  watch_points: string[]
  metrics: MetricCard[]
}

export interface StockItem {
  symbol: string
  name: string
  board: string
  industry: string
  latest_price: number
  change_pct: number
  turnover_ratio: number
  pe_ttm: number
  market_cap: number
  score: number
  thesis: string
  tags: string[]
  in_watchlist: boolean
}

export interface StockListResponse {
  total: number
  rows: StockItem[]
}

export interface PricePoint {
  date: string
  open: number
  close: number
  low: number
  high: number
  volume: number
  ma5: number
  ma20: number
}

export interface SignalBreakdown {
  dimension: string
  score: number
  takeaway: string
}

export interface RecommendationConfidenceSignal {
  dimension: string
  score: number
  takeaway: string
}

export interface RecommendationTrustSummary {
  data_mode: 'demo' | 'live'
  snapshot_updated_at: string
  strongest_signals: RecommendationConfidenceSignal[]
  primary_risk: string
  confidence_score: number
  confidence_notice: string
}

export type MoveBias = 'bullish' | 'mixed' | 'cautious'

export interface MoveDriver {
  title: string
  detail: string
  strength: number
  tone: 'positive' | 'negative'
}

export interface MoveAnalysis {
  bias: MoveBias
  summary: string
  positive_drivers: MoveDriver[]
  negative_drivers: MoveDriver[]
  watch_points: string[]
}

export type EventTone = 'positive' | 'neutral' | 'caution'

export interface EventItem {
  date: string | null
  category: string
  title: string
  headline: string
  detail: string
  tone: EventTone
  source: string
  url: string | null
}

export interface EventAnalysis {
  tone: EventTone
  summary: string
  tags: string[]
  items: EventItem[]
  watch_points: string[]
}

export type CapitalFlowTone = 'positive' | 'neutral' | 'caution'

export interface CapitalFlowAnalysis {
  status: 'ready' | 'derived' | 'placeholder'
  tone: CapitalFlowTone
  summary: string
  latest_trade_date: string | null
  main_net_inflow_1d: number | null
  main_net_ratio_1d: number | null
  main_net_inflow_5d: number | null
  active_days_5d: number | null
  ultra_large_net_inflow_1d: number | null
  lhb_on_list_count: number | null
  lhb_recent_date: string | null
  lhb_net_buy_amount: number | null
  watch_points: string[]
}

export interface RecommendationDiagnosis {
  is_recommended: boolean
  current_rank: number
  total_candidates: number
  recommendation_limit: number
  gap_to_limit: number
  score_gap_to_cutoff: number | null
  cutoff_symbol: string | null
  cutoff_name: string | null
  cutoff_score: number | null
  summary: string
  reason_points: string[]
  blocking_points: string[]
  action_points: string[]
}

export interface FundamentalSnapshot {
  report_period: string | null
  revenue_growth: number | null
  net_profit_growth: number | null
  deduct_profit_growth: number | null
  roe: number | null
  gross_margin: number | null
  debt_ratio: number | null
  eps: number | null
  operating_cashflow_per_share: number | null
}

export interface StockDetail extends StockItem {
  price_series: PricePoint[]
  thesis_points: string[]
  risk_notes: string[]
  signal_breakdown: SignalBreakdown[]
  fundamental: FundamentalSnapshot | null
  move_analysis: MoveAnalysis | null
  event_analysis: EventAnalysis | null
  capital_flow_analysis: CapitalFlowAnalysis | null
  recommendation_diagnosis: RecommendationDiagnosis | null
  recommendation_trust: RecommendationTrustSummary | null
}

export interface RecommendationItem {
  symbol: string
  name: string
  score: number
  entry_window: string
  expected_holding_days: number
  thesis: string
  risk: string
  tags: string[]
  latest_price: number | null
  recent_return_5d: number | null
  recent_return_20d: number | null
  move_bias: MoveBias | null
  move_summary: string | null
  event_tone: EventTone | null
  event_summary: string | null
  data_mode: 'demo' | 'live'
  snapshot_updated_at: string
  strongest_signals: RecommendationConfidenceSignal[]
  primary_risk: string
  confidence_score: number
  confidence_notice: string
  in_watchlist: boolean
}

export interface RecommendationJournalItem {
  run_key: string
  generated_at: string
  symbol: string
  name: string
  score: number
  entry_window: string
  expected_holding_days: number
  thesis: string
  risk: string
  source: string
  data_mode: 'demo' | 'live'
  tags: string[]
  price_at_publish: number
  current_price: number | null
  current_return: number | null
  days_since_publish: number
  tracking_status: 'tracking' | 'matured'
  is_matured_for_expected_window: boolean
}

export interface RecommendationReviewWindowMetric {
  window_days: number
  sample_size: number
  win_rate: number | null
  avg_return: number | null
  best_return: number | null
  worst_return: number | null
}

export interface RecommendationReviewRun {
  run_key: string
  generated_at: string
  source: string
  picks: number
  avg_score: number
  avg_return_5d: number | null
  avg_return_10d: number | null
  avg_return_20d: number | null
  avg_expected_return: number | null
}

export interface RecommendationReviewSample {
  run_key: string
  generated_at: string
  symbol: string
  name: string
  score: number
  source: string
  data_mode: 'demo' | 'live'
  entry_window: string
  expected_holding_days: number
  thesis: string
  tags: string[]
  price_at_publish: number
  latest_known_price: number | null
  return_5d: number | null
  return_10d: number | null
  return_20d: number | null
  expected_return: number | null
}

export interface RecommendationModeBreakdown {
  mode: 'demo' | 'live'
  sample_size: number
}

export interface RecommendationReviewMaturityMetric {
  window_days: number
  total_samples: number
  matured_samples: number
  immature_samples: number
}

export interface RecommendationReviewResponse {
  total_samples: number
  evaluation_mode: 'demo' | 'live'
  evaluation_notice: string
  trust_level: 'low' | 'medium' | 'high'
  trust_reasons: string[]
  mode_breakdown: RecommendationModeBreakdown[]
  maturity_breakdown: RecommendationReviewMaturityMetric[]
  window_metrics: RecommendationReviewWindowMetric[]
  recent_runs: RecommendationReviewRun[]
  top_hits: RecommendationReviewSample[]
  top_misses: RecommendationReviewSample[]
  samples: RecommendationReviewSample[]
}

export interface StrategyConfig {
  technical_weight: number
  fundamental_weight: number
  money_flow_weight: number
  sentiment_weight: number
  rebalance_cycle: 'daily' | 'weekly' | 'biweekly'
  min_turnover: number
  min_listing_days: number
  exclude_st: boolean
  exclude_new_shares: boolean
}

export interface SyncTask {
  task_key: string
  name: string
  status: 'idle' | 'running' | 'success' | 'warning'
  schedule: string
  last_run_at: string | null
  next_run_at: string | null
  message: string
  source: string
}

export interface DataSourceStatusItem {
  provider_key: string
  display_name: string
  enabled: boolean
  priority: number
  supports_snapshot: boolean
  supports_history: boolean
  supports_fundamental: boolean
  last_status: 'idle' | 'success' | 'warning'
  last_message: string
  last_run_at: string | null
  last_success_at: string | null
  last_failure_at: string | null
  updated_at: string
}

export interface EventSyncOverview {
  status: 'idle' | 'partial' | 'placeholder' | 'ready'
  summary: string
  configured_sources: string[]
  detected_sources: string[]
  coverage_count: number
  total_symbols: number
  active_symbols: number
  total_items: number
  updated_at: string | null
}

export interface DataSourceOverview {
  current_provider: string
  fallback_chain: string[]
  items: DataSourceStatusItem[]
  event_sync: EventSyncOverview
}

export type WatchlistStatus = 'watching' | 'holding' | 'archived'

export interface WatchlistCreateRequest {
  symbol: string
  source?: 'manual' | 'recommendation'
  status?: WatchlistStatus
  notes?: string | null
}

export interface WatchlistUpdateRequest {
  status?: WatchlistStatus
  notes?: string | null
}

export interface WatchlistItem {
  symbol: string
  name: string
  board: string | null
  industry: string | null
  status: WatchlistStatus
  source: string
  notes: string | null
  added_price: number | null
  latest_price: number | null
  change_pct: number | null
  score: number | null
  thesis: string | null
  tags: string[]
  current_return: number | null
  added_at: string
  updated_at: string
}

export type TradePlanStatus = 'planned' | 'active' | 'closed' | 'cancelled'
export type TradePlanSource = 'manual' | 'recommendation' | 'watchlist'

export interface TradePlanCreateRequest {
  symbol: string
  source?: TradePlanSource
  status?: TradePlanStatus
  thesis?: string | null
  notes?: string | null
  planned_entry_price?: number | null
  actual_entry_price?: number | null
  actual_exit_price?: number | null
  stop_loss_price?: number | null
  target_price?: number | null
  planned_position_pct?: number | null
}

export interface TradePlanUpdateRequest {
  status?: TradePlanStatus
  thesis?: string | null
  notes?: string | null
  planned_entry_price?: number | null
  actual_entry_price?: number | null
  actual_exit_price?: number | null
  stop_loss_price?: number | null
  target_price?: number | null
  planned_position_pct?: number | null
}

export interface TradePlanItem {
  id: number
  symbol: string
  name: string
  board: string | null
  industry: string | null
  source: TradePlanSource
  status: TradePlanStatus
  thesis: string | null
  notes: string | null
  planned_entry_price: number | null
  actual_entry_price: number | null
  actual_exit_price: number | null
  stop_loss_price: number | null
  target_price: number | null
  planned_position_pct: number | null
  latest_price: number | null
  change_pct: number | null
  score: number | null
  tags: string[]
  plan_gap_pct: number | null
  current_return: number | null
  realized_return: number | null
  risk_reward_ratio: number | null
  created_at: string
  opened_at: string | null
  closed_at: string | null
  updated_at: string
}

export type PortfolioPositionStatus = 'holding' | 'closed'
export type PortfolioPositionSource = 'manual' | 'trade_plan' | 'recommendation' | 'watchlist'
export type PortfolioRiskLevel = 'low' | 'medium' | 'high'

export interface PortfolioProfileConfig {
  name: string
  initial_capital: number
  benchmark: string
  notes: string | null
}

export interface PortfolioPositionCreateRequest {
  symbol: string
  source?: PortfolioPositionSource
  status?: PortfolioPositionStatus
  quantity: number
  entry_price: number
  exit_price?: number | null
  stop_loss_price?: number | null
  target_price?: number | null
  thesis?: string | null
  notes?: string | null
}

export interface PortfolioPositionUpdateRequest {
  status?: PortfolioPositionStatus
  quantity?: number | null
  entry_price?: number | null
  exit_price?: number | null
  stop_loss_price?: number | null
  target_price?: number | null
  thesis?: string | null
  notes?: string | null
}

export interface PortfolioPositionItem {
  id: number
  symbol: string
  name: string
  board: string | null
  industry: string | null
  source: PortfolioPositionSource
  status: PortfolioPositionStatus
  quantity: number
  entry_price: number
  exit_price: number | null
  stop_loss_price: number | null
  target_price: number | null
  latest_price: number | null
  change_pct: number | null
  score: number | null
  tags: string[]
  thesis: string | null
  notes: string | null
  cost_value: number
  market_value: number | null
  unrealized_pnl: number | null
  unrealized_return: number | null
  realized_pnl: number | null
  realized_return: number | null
  weight_pct: number | null
  risk_level: PortfolioRiskLevel
  risk_flags: string[]
  stop_distance_pct: number | null
  target_distance_pct: number | null
  created_at: string
  opened_at: string | null
  closed_at: string | null
  updated_at: string
}

export interface PortfolioIndustryExposure {
  industry: string
  market_value: number
  weight_pct: number
}

export interface PortfolioSummary {
  initial_capital: number
  estimated_cash: number
  estimated_total_assets: number
  invested_cost: number
  market_value: number
  holding_count: number
  closed_count: number
  unrealized_pnl: number
  realized_pnl: number
  total_return_pct: number
  utilization_pct: number
  winning_count: number
  losing_count: number
  at_risk_position_count: number
  capital_at_risk: number
  capital_at_risk_pct: number
  top_industry: string | null
  top_industry_weight_pct: number | null
  worst_position_name: string | null
  worst_position_return_pct: number | null
  risk_level: PortfolioRiskLevel
  largest_weight_pct: number | null
}

export interface PortfolioOverview {
  profile: PortfolioProfileConfig
  summary: PortfolioSummary
  industry_exposure: PortfolioIndustryExposure[]
  positions: PortfolioPositionItem[]
}

export type AlertStatus = 'active' | 'handled' | 'resolved'
export type AlertSeverity = 'critical' | 'warning' | 'info'
export type AlertCategory = 'trade_plan' | 'portfolio' | 'watchlist'

export interface AlertItem {
  id: number
  event_key: string
  status: AlertStatus
  severity: AlertSeverity
  category: AlertCategory
  kind: string
  symbol: string | null
  name: string | null
  title: string
  message: string
  action_path: string | null
  source_type: string | null
  source_id: number | null
  last_value: number | null
  threshold_value: number | null
  payload: Record<string, unknown>
  created_at: string
  last_seen_at: string
  resolved_at: string | null
  updated_at: string
}

export interface AlertOverview {
  total_count: number
  filtered_count: number
  active_count: number
  handled_count: number
  resolved_count: number
  critical_count: number
  warning_count: number
  info_count: number
  latest_evaluated_at: string | null
  items: AlertItem[]
}

File: backend/app/schemas/market.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

WatchlistStatus = Literal["watching", "holding", "archived"]
TradePlanStatus = Literal["planned", "active", "closed", "cancelled"]
TradePlanSource = Literal["manual", "recommendation", "watchlist"]
PortfolioPositionStatus = Literal["holding", "closed"]
PortfolioPositionSource = Literal["manual", "trade_plan", "recommendation", "watchlist"]
PortfolioRiskLevel = Literal["low", "medium", "high"]
AlertStatus = Literal["active", "handled", "resolved"]
AlertSeverity = Literal["critical", "warning", "info"]
AlertCategory = Literal["trade_plan", "portfolio", "watchlist"]
MoveBias = Literal["bullish", "mixed", "cautious"]
EventTone = Literal["positive", "neutral", "caution"]
CapitalFlowTone = Literal["positive", "neutral", "caution"]


class HealthResponse(BaseModel):
    status: str
    app_name: str
    mode: str
    scheduler_enabled: bool
    market_db_path: str
    strategy_store: str


class MetricCard(BaseModel):
    label: str
    value: str
    change: str
    tone: Literal["positive", "negative", "neutral"]
    description: str


class IndustryHeat(BaseModel):
    industry: str
    score: int
    momentum: str


class MarketPulsePoint(BaseModel):
    date: str
    score: int
    turnover: float


class RecommendationItem(BaseModel):
    symbol: str
    name: str
    score: int
    entry_window: str
    expected_holding_days: int
    thesis: str
    risk: str
    tags: list[str]
    latest_price: float | None = None
    recent_return_5d: float | None = None
    recent_return_20d: float | None = None
    move_bias: MoveBias | None = None
    move_summary: str | None = None
    event_tone: EventTone | None = None
    event_summary: str | None = None
    in_watchlist: bool = False


class RecommendationJournalItem(BaseModel):
    run_key: str
    generated_at: str
    symbol: str
    name: str
    score: int
    entry_window: str
    expected_holding_days: int
    thesis: str
    risk: str
    source: str
    tags: list[str]
    price_at_publish: float
    current_price: float | None = None
    current_return: float | None = None


class RecommendationReviewWindowMetric(BaseModel):
    window_days: int
    sample_size: int
    win_rate: float | None = None
    avg_return: float | None = None
    best_return: float | None = None
    worst_return: float | None = None


class RecommendationReviewRun(BaseModel):
    run_key: str
    generated_at: str
    source: str
    picks: int
    avg_score: float
    avg_return_5d: float | None = None
    avg_return_10d: float | None = None
    avg_return_20d: float | None = None
    avg_expected_return: float | None = None


class RecommendationReviewSample(BaseModel):
    run_key: str
    generated_at: str
    symbol: str
    name: str
    score: int
    source: str
    entry_window: str
    expected_holding_days: int
    thesis: str
    tags: list[str]
    price_at_publish: float
    latest_known_price: float | None = None
    return_5d: float | None = None
    return_10d: float | None = None
    return_20d: float | None = None
    expected_return: float | None = None


class RecommendationReviewResponse(BaseModel):
    total_samples: int
    window_metrics: list[RecommendationReviewWindowMetric]
    recent_runs: list[RecommendationReviewRun]
    top_hits: list[RecommendationReviewSample]
    top_misses: list[RecommendationReviewSample]
    samples: list[RecommendationReviewSample]


class MarketContextSummary(BaseModel):
    regime: Literal["risk_on", "balanced", "risk_off"]
    regime_label: str
    summary: str
    action_hint: str
    watch_points: list[str]
    metrics: list[MetricCard]


class BenchmarkIndex(BaseModel):
    code: str
    name: str
    latest_price: float
    change_pct: float
    return_20d: float | None = None
    trend: str
    takeaway: str


class MarketBreadthSnapshot(BaseModel):
    scope_label: str
    total_count: int
    advancers: int
    decliners: int
    advance_ratio: float
    strong_count: int
    strong_ratio: float
    avg_change: float
    avg_turnover: float
    top_industry: str | None = None
    top_two_share: float
    limit_up_like: int | None = None
    limit_down_like: int | None = None
    summary: str


class MarketCapitalFlowOverview(BaseModel):
    status: Literal["ready", "derived", "placeholder"]
    scope_label: str
    summary: str
    watch_points: list[str]
    metrics: list[MetricCard]


class DashboardSummary(BaseModel):
    headline: str
    updated_at: str
    market_context: MarketContextSummary
    benchmark_indices: list[BenchmarkIndex]
    breadth_snapshot: MarketBreadthSnapshot
    market_capital_flow: MarketCapitalFlowOverview
    market_overview: list[MetricCard]
    hot_industries: list[IndustryHeat]
    market_pulse: list[MarketPulsePoint]
    top_recommendations: list[RecommendationItem]
    risk_flags: list[str]


class StockItem(BaseModel):
    symbol: str
    name: str
    board: str
    industry: str
    latest_price: float
    change_pct: float
    turnover_ratio: float
    pe_ttm: float
    market_cap: float
    score: int
    thesis: str
    tags: list[str]
    in_watchlist: bool = False


class StockListResponse(BaseModel):
    total: int
    rows: list[StockItem]


class PricePoint(BaseModel):
    date: str
    open: float
    close: float
    low: float
    high: float
    volume: float
    ma5: float
    ma20: float


class SignalBreakdown(BaseModel):
    dimension: str
    score: int
    takeaway: str


class FundamentalSnapshot(BaseModel):
    report_period: str | None = None
    revenue_growth: float | None = None
    net_profit_growth: float | None = None
    deduct_profit_growth: float | None = None
    roe: float | None = None
    gross_margin: float | None = None
    debt_ratio: float | None = None
    eps: float | None = None
    operating_cashflow_per_share: float | None = None


class MoveDriver(BaseModel):
    title: str
    detail: str
    strength: int
    tone: Literal["positive", "negative"]


class MoveAnalysis(BaseModel):
    bias: MoveBias
    summary: str
    positive_drivers: list[MoveDriver]
    negative_drivers: list[MoveDriver]
    watch_points: list[str]


class EventItem(BaseModel):
    date: str | None = None
    category: str
    title: str
    headline: str
    detail: str
    tone: EventTone
    source: str
    url: str | None = None


class EventAnalysis(BaseModel):
    tone: EventTone
    summary: str
    tags: list[str]
    items: list[EventItem]
    watch_points: list[str]


class CapitalFlowAnalysis(BaseModel):
    status: Literal["ready", "derived", "placeholder"]
    tone: CapitalFlowTone
    summary: str
    latest_trade_date: str | None = None
    main_net_inflow_1d: float | None = None
    main_net_ratio_1d: float | None = None
    main_net_inflow_5d: float | None = None
    active_days_5d: int | None = None
    ultra_large_net_inflow_1d: float | None = None
    lhb_on_list_count: int | None = None
    lhb_recent_date: str | None = None
    lhb_net_buy_amount: float | None = None
    watch_points: list[str]


class RecommendationDiagnosis(BaseModel):
    is_recommended: bool
    current_rank: int
    total_candidates: int
    recommendation_limit: int
    gap_to_limit: int
    score_gap_to_cutoff: int | None = None
    cutoff_symbol: str | None = None
    cutoff_name: str | None = None
    cutoff_score: int | None = None
    summary: str
    reason_points: list[str]
    blocking_points: list[str]
    action_points: list[str]


class StockDetail(StockItem):
    price_series: list[PricePoint]
    thesis_points: list[str]
    risk_notes: list[str]
    signal_breakdown: list[SignalBreakdown]
    fundamental: FundamentalSnapshot | None = None
    move_analysis: MoveAnalysis | None = None
    event_analysis: EventAnalysis | None = None
    capital_flow_analysis: CapitalFlowAnalysis | None = None
    recommendation_diagnosis: RecommendationDiagnosis | None = None


class StrategyConfig(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    technical_weight: int
    fundamental_weight: int
    money_flow_weight: int
    sentiment_weight: int
    rebalance_cycle: Literal["daily", "weekly", "biweekly"]
    min_turnover: float
    min_listing_days: int
    exclude_st: bool
    exclude_new_shares: bool


class SyncTask(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    task_key: str
    name: str
    status: Literal["idle", "running", "success", "warning"]
    schedule: str
    last_run_at: str | None
    next_run_at: str | None
    message: str
    source: str


class DataSourceStatusItem(BaseModel):
    provider_key: str
    display_name: str
    enabled: bool
    priority: int
    supports_snapshot: bool
    supports_history: bool
    supports_fundamental: bool
    last_status: Literal["idle", "success", "warning"]
    last_message: str
    last_run_at: str | None = None
    last_success_at: str | None = None
    last_failure_at: str | None = None
    updated_at: str


class EventSyncOverview(BaseModel):
    status: Literal["idle", "partial", "placeholder", "ready"]
    summary: str
    configured_sources: list[str]
    detected_sources: list[str]
    coverage_count: int
    total_symbols: int
    active_symbols: int
    total_items: int
    updated_at: str | None = None


class DataSourceOverview(BaseModel):
    current_provider: str
    fallback_chain: list[str]
    items: list[DataSourceStatusItem]
    event_sync: EventSyncOverview


class WatchlistCreateRequest(BaseModel):
    symbol: str
    source: Literal["manual", "recommendation"] = "manual"
    status: WatchlistStatus = "watching"
    notes: str | None = None


class WatchlistUpdateRequest(BaseModel):
    status: WatchlistStatus | None = None
    notes: str | None = None


class WatchlistItem(BaseModel):
    symbol: str
    name: str
    board: str | None = None
    industry: str | None = None
    status: WatchlistStatus
    source: str
    notes: str | None = None
    added_price: float | None = None
    latest_price: float | None = None
    change_pct: float | None = None
    score: int | None = None
    thesis: str | None = None
    tags: list[str]
    current_return: float | None = None
    added_at: str
    updated_at: str


class TradePlanCreateRequest(BaseModel):
    symbol: str
    source: TradePlanSource = "manual"
    status: TradePlanStatus = "planned"
    thesis: str | None = None
    notes: str | None = None
    planned_entry_price: float | None = Field(default=None, ge=0)
    actual_entry_price: float | None = Field(default=None, ge=0)
    actual_exit_price: float | None = Field(default=None, ge=0)
    stop_loss_price: float | None = Field(default=None, ge=0)
    target_price: float | None = Field(default=None, ge=0)
    planned_position_pct: int | None = Field(default=None, ge=1, le=100)


class TradePlanUpdateRequest(BaseModel):
    status: TradePlanStatus | None = None
    thesis: str | None = None
    notes: str | None = None
    planned_entry_price: float | None = Field(default=None, ge=0)
    actual_entry_price: float | None = Field(default=None, ge=0)
    actual_exit_price: float | None = Field(default=None, ge=0)
    stop_loss_price: float | None = Field(default=None, ge=0)
    target_price: float | None = Field(default=None, ge=0)
    planned_position_pct: int | None = Field(default=None, ge=1, le=100)


class TradePlanItem(BaseModel):
    id: int
    symbol: str
    name: str
    board: str | None = None
    industry: str | None = None
    source: TradePlanSource
    status: TradePlanStatus
    thesis: str | None = None
    notes: str | None = None
    planned_entry_price: float | None = None
    actual_entry_price: float | None = None
    actual_exit_price: float | None = None
    stop_loss_price: float | None = None
    target_price: float | None = None
    planned_position_pct: int | None = None
    latest_price: float | None = None
    change_pct: float | None = None
    score: int | None = None
    tags: list[str]
    plan_gap_pct: float | None = None
    current_return: float | None = None
    realized_return: float | None = None
    risk_reward_ratio: float | None = None
    created_at: str
    opened_at: str | None = None
    closed_at: str | None = None
    updated_at: str


class PortfolioProfileConfig(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    initial_capital: float = Field(ge=1000)
    benchmark: str
    notes: str | None = None


class PortfolioPositionCreateRequest(BaseModel):
    symbol: str
    source: PortfolioPositionSource = "manual"
    status: PortfolioPositionStatus = "holding"
    quantity: int = Field(ge=1)
    entry_price: float = Field(ge=0)
    exit_price: float | None = Field(default=None, ge=0)
    stop_loss_price: float | None = Field(default=None, ge=0)
    target_price: float | None = Field(default=None, ge=0)
    thesis: str | None = None
    notes: str | None = None


class PortfolioPositionUpdateRequest(BaseModel):
    status: PortfolioPositionStatus | None = None
    quantity: int | None = Field(default=None, ge=1)
    entry_price: float | None = Field(default=None, ge=0)
    exit_price: float | None = Field(default=None, ge=0)
    stop_loss_price: float | None = Field(default=None, ge=0)
    target_price: float | None = Field(default=None, ge=0)
    thesis: str | None = None
    notes: str | None = None


class PortfolioPositionItem(BaseModel):
    id: int
    symbol: str
    name: str
    board: str | None = None
    industry: str | None = None
    source: PortfolioPositionSource
    status: PortfolioPositionStatus
    quantity: int
    entry_price: float
    exit_price: float | None = None
    stop_loss_price: float | None = None
    target_price: float | None = None
    latest_price: float | None = None
    change_pct: float | None = None
    score: int | None = None
    tags: list[str]
    thesis: str | None = None
    notes: str | None = None
    cost_value: float
    market_value: float | None = None
    unrealized_pnl: float | None = None
    unrealized_return: float | None = None
    realized_pnl: float | None = None
    realized_return: float | None = None
    weight_pct: float | None = None
    risk_level: PortfolioRiskLevel
    risk_flags: list[str]
    stop_distance_pct: float | None = None
    target_distance_pct: float | None = None
    created_at: str
    opened_at: str | None = None
    closed_at: str | None = None
    updated_at: str


class PortfolioIndustryExposure(BaseModel):
    industry: str
    market_value: float
    weight_pct: float


class PortfolioSummary(BaseModel):
    initial_capital: float
    estimated_cash: float
    estimated_total_assets: float
    invested_cost: float
    market_value: float
    holding_count: int
    closed_count: int
    unrealized_pnl: float
    realized_pnl: float
    total_return_pct: float
    utilization_pct: float
    winning_count: int
    losing_count: int
    at_risk_position_count: int
    capital_at_risk: float
    capital_at_risk_pct: float
    top_industry: str | None = None
    top_industry_weight_pct: float | None = None
    worst_position_name: str | None = None
    worst_position_return_pct: float | None = None
    risk_level: PortfolioRiskLevel
    largest_weight_pct: float | None = None


class PortfolioOverview(BaseModel):
    profile: PortfolioProfileConfig
    summary: PortfolioSummary
    industry_exposure: list[PortfolioIndustryExposure]
    positions: list[PortfolioPositionItem]


class AlertStatusUpdateRequest(BaseModel):
    status: AlertStatus


class AlertItem(BaseModel):
    id: int
    event_key: str
    status: AlertStatus
    severity: AlertSeverity
    category: AlertCategory
    kind: str
    symbol: str | None = None
    name: str | None = None
    title: str
    message: str
    action_path: str | None = None
    source_type: str | None = None
    source_id: int | None = None
    last_value: float | None = None
    threshold_value: float | None = None
    payload: dict[str, object] = Field(default_factory=dict)
    created_at: str
    last_seen_at: str
    resolved_at: str | None = None
    updated_at: str


class AlertOverview(BaseModel):
    total_count: int
    filtered_count: int
    active_count: int
    handled_count: int
    resolved_count: int
    critical_count: int
    warning_count: int
    info_count: int
    latest_evaluated_at: str | None = None
    items: list[AlertItem]
